src/query/xasl_to_stream.c

```diff
@@ -1454,6 +1454,81 @@ xts_save_val_list (const VAL_LIST * val_list)
   return offset;
 }
 
+static char *
+xts_process_cte_xasl_id (char *ptr, const XASL_ID * cte_xasl_id)
+{
+  int i;
+
+  for (i = 0; i < 5; i++)
+    {
+      OR_PUT_INT (ptr, cte_xasl_id->sha1.h[i]);
+      ptr += OR_INT_SIZE;
+    }
+
+  OR_PUT_INT (ptr, cte_xasl_id->cache_flag);
+  ptr += OR_INT_SIZE;
+  OR_PUT_INT (ptr, cte_xasl_id->time_stored.sec);
+  ptr += OR_INT_SIZE;
+  OR_PUT_INT (ptr, cte_xasl_id->time_stored.usec);
+  ptr += OR_INT_SIZE;
+
+  return ptr;
+}
+
+static int
+xts_save_cte_xasl_id (const XASL_ID * cte_xasl_id)
+{
+  int offset, i;
+  int size, nelements = 8;
+  char *ptr;
+
+  OR_ALIGNED_BUF (sizeof (*cte_xasl_id)) a_buf;
+  char *buf = OR_ALIGNED_BUF_START (a_buf);
+
+  char *buf_p = NULL;
+  bool is_buf_alloced = false;
+
+  if (cte_xasl_id == NULL)
+    {
+      return NO_ERROR;
+    }
+
+  offset = xts_get_offset_visited_ptr (cte_xasl_id);
+  if (offset != ER_FAILED)
+    {
+      return offset;
+    }
+
+  size = OR_INT_SIZE * nelements;
+
+  offset = xts_reserve_location_in_stream (size);
+  if (offset == ER_FAILED || xts_mark_ptr_visited (cte_xasl_id, offset) == ER_FAILED)
+    {
+      return ER_FAILED;
+    }
+
+  if (size <= (int) OR_ALIGNED_BUF_SIZE (a_buf))
+    {
+      buf_p = buf;
+    }
+  else
+    {
+      buf_p = (char *) malloc (size);
+      if (buf_p == NULL)
+	{
+	  xts_Xasl_errcode = ER_OUT_OF_VIRTUAL_MEMORY;
+	  return ER_FAILED;
+	}
+
+      is_buf_alloced = true;
+    }
+
+  buf = xts_process_cte_xasl_id (buf_p, cte_xasl_id);
+  memcpy (&xts_Stream_buffer[offset], buf_p, size);
+
```

Comment:
You need to free buf_p.
Also, Is there any guarantee that the space in xts_Stream_buffer[offset] is large enough?

Reply:
The code is resemble with xts_process_cte_list_id.